[WIP] avoid unnecessary calls to get_cursor_line_ptr

JackDanger opened the pull request:
This is my first contribution and I figured I'd clean up `skip_label`.

This commit refactors the function-local variables to be initialized or
reassigned a fewer number of times.

JackDanger committed:
refactor indent_c's skip_label

This commit refactors the function-local variables to be initialized or
reassigned a fewer number of times. It calls `ml_get(lnum)` directly
rather than relying on `get_cursor_line_ptr()` and doesn't attempt to
store, modify, and restore the current window's w_cursor field.

justinmk commented:
Thanks! Would you mind signing the CLA? See CONTRIBUTING.md.

aktau commented:
It's likely possible to simplify this a bit more. I'll write something up.

aktau commented:
Looking at this, maybe a slightly more invasive refactoring is in order. Here are some of the (direct and indirect) helper functions that are being used:

```c
char_u *get_cursor_line_ptr(void)
{
  return ml_get_buf(curbuf, curwin->w_cursor.lnum, false);
}

char_u *ml_get(linenr_T lnum)
{
  return ml_get_buf(curbuf, lnum, FALSE);
}

// Count the size (in window cells) of the indent in the current line.
int get_indent(void)
{
  return get_indent_str(get_cursor_line_ptr(), (int)curbuf->b_p_ts, false);
}
```

So it appears to me that `get_cursor_line_ptr()` can be replaced by `ml_get(lnum)` without ill effects (it's also shorter). I know what `ml_get()` does since it's such an oft-used function, yet it wasn't sure what `get_cursor_line_ptr()` did and I'm not sure if I'll know if I look at it next month.

I think that if we do that, we also won't need to save and restore the current cursor anymore. This was probably done to make functions like `get_cursor_line_ptr()` and some others work correctly. In the version below, I've replaced these context-aware functions with functions that take parameters.

This also shortens the function quite a bit. This is my current version:

```c
// Find indent for line "lnum", ignoring any case or jump label.
// Also return a pointer to the text (after the label) in "pp".
//   label:	if (asdf && asdfasdf)
//		^
static int skip_label(linenr_T lnum, char_u **pp) FUNC_ATTR_NONNULL_ALL
{
  char_u *l = ml_get(lnum);

  if (cin_iscase(l, FALSE) || cin_isscopedecl(l) || cin_islabel()) {
    char_u *i = after_label(l);
    *pp = (i == NULL) ? l : i;
    return get_indent_nolabel(lnum);
  }

  *pp = l;
  return get_indent_str(l, (int)curbuf->b_p_ts, false);
}
```

Note that this is completely untested. So that's something nice to test :).

JackDanger commented:
@justinmk Thanks so much! I've signed the CLA and I'll take another look at this today. I appreciate you helping me get up to speed with contributing in an unfamiliar codebase. I'll likely be a bit slow as I figure out which pieces of this code are critical and which aren't but I'm really happy to have something to work on.

JackDanger commented:
@aktau Thanks for giving me this direction. I'm looking into applying your ideas right now.

JackDanger commented:
@aktau I've applied your suggestions (with some small modifications where necessary). I also renamed the function to be much clearer about what it does.

I tried to find a place to put a test for something like this but I came up empty. Can you point me to a similar-ish test that I can use as a reference?

JackDanger commented:
Could somebody who's able to run the tests successfully on master run them on this branch to see why the Travis build is failing?

I'm having trouble running the tests (from both master and this branch on both Mac and Linux) due to some kind of fork resource failure.

oni-link commented on src/nvim/indent_c.c:
```diff
@@ -508,29 +508,21 @@ get_indent_nolabel (     /* XXX */
+    // calls it).
+    char_u *i = after_label(ml_get(lnum));
+    *pp = (i == NULL) ? (char_u *)1 : i;
+    return get_indent_nolabel(lnum);
```
`get_indent_nolabel(lnum)` also calls `ml_get(lnum)`, so at this point `*pp` is an invalid pointer.

oni-link commented on src/nvim/indent_c.c:
```diff
@@ -508,29 +508,21 @@ get_indent_nolabel (     /* XXX */
+    // ml_get invalidates the results of previous calls (and "cin_islabel()"
+    // calls it).
+    char_u *i = after_label(ml_get(lnum));
+    *pp = (i == NULL) ? (char_u *)1 : i;
```
Instead of `(char_u *)1` we need `ml_get(lnum)` here.

oni-link commented on src/nvim/indent_c.c:
```diff
@@ -508,29 +508,21 @@ get_indent_nolabel (     /* XXX */
 
-  curwin->w_cursor = cursor_save;
-  return amount;
+  *pp = l;
```
`l` can be invalid at this point, because the previous `if` statement could have used `cin_islabel()`.

JackDanger committed:
Updating names of label-ignoring indent methods

Rename: "get_indent_nolabel" => "get_indent_after_label_with_ref"
Rename: "skip_label" => "get_indent_after_label_p"

JackDanger commented:
@oni-link I updated this code to be much smarter about how any return value of `ml_get` is used. I avoided calls to `ml_get` when possible but made sure the code never references an invalidated pointer.

I also changed the names of `skip_label` and `get_indent_nolabel` to make more sense. They are now `get_indent_after_label` and `get_indent_after_label_pp` for the normal form and the form that returns the pointer, respectively.

How does it look to you?

oni-link commented on src/nvim/indent_c.c:
```diff
@@ -481,15 +481,13 @@ static char_u *after_label(char_u *l)
-{
-  char_u      *l;
+static int get_indent_after_label(linenr_T lnum)
+  FUNC_ATTR_NONNULL_ALL
```
No need for this attribute. We don't have any pointer arguments.

JackDanger replied:
Got it, I misread the docs for these attributes. Thanks.

oni-link commented on src/nvim/indent_c.c:
```diff
@@ -508,29 +506,25 @@ get_indent_nolabel (     /* XXX */
-    if (l == NULL)              /* just in case */
-      l = get_cursor_line_ptr();
+    indent = get_indent_after_label(lnum);
+    char_u *i = after_label(ml_get(lnum));
```
Why not reuse variable `l` here?

JackDanger replied:
Because `get_indent_after_label` calls `ml_get_*` stuff and invalidates `l` :(
I tried to organize things to reuse pointers when possible but there are many, many functions that call `ml_get`.

oni-link replied:
Reuse for the result, not for the argument. Why introduce a new variable?

JackDanger replied:
Ah, I misunderstood, sorry. Yeah, why did I introduce a new variable? Fixing.

aktau commented:
There you have it, I didn't even think about the fact that `ml_get()` apparently invalidates former pointers. I can leave this in the capable hands of @oni-link thus :). Keep up the good work @JackDanger!

oni-link commented on src/nvim/indent_c.c:
```diff
@@ -508,29 +505,25 @@ get_indent_nolabel (     /* XXX */
-  /* XXX */
+  // We repeatedly call "ml_get(lnum)" because every time it's called (or any
+  // time we call a function that calls it, marked with an 'XXX' comment) the
+  // returned pointer is invalidated.
```
If we update the function comment to the new format (and replace tabs with spaces), it could be helpful to have a note or warning about this. Perhaps something like
`/// @warning This function invalidates pointers obtained with the ml_get_*() functions`.

JackDanger replied:
I didn't find any tabs here to change to spaces, can you tell me which tabs you mean?
Weirdly, the convention throughout the file is to add the `/* XXX */` comment whenever you might invalid an `ml_get_*()` pointer. It took me forever to realize this though. I've added one of those comments to this function's definition and also inline with the `cin_islabel()` call (which is the thing on line 517 that calls ml_get). How's it look?

oni-link replied:
We started to change the documentation style for functions (doxygen). New functions always follow this style and old ones could be updated. Have a look at memory.c for examples.  

> I didn't find any tabs here to change to spaces, can you tell me which tabs you mean?

The function description has an example, which contains tabs:
```c
/*
 * Find indent for line "lnum", ignoring any case or jump label.
 * Also return a pointer to the text (after the label) in "pp".
 *   label:	if (asdf && asdfasdf)
 *		^
 */
```
The arrow (`^`) is not aligned with the `if` for me, because tabs are used for indentation and their size depends on option `'tabstop'`.


> Weirdly, the convention throughout the file is to add the /* XXX */ comment whenever you might invalid an ml_get_*() pointer. It took me forever to realize this though.

In my opinion, with a note or warning in the function description about the invalidation of pointers, the meaning of `/* XXX */` is easier to get and better documented. But I would leave the `/* XXX */` in the code until all affected function would have been updated.

JackDanger replied:
> The arrow (^) is not aligned with the if for me, because tabs are used for indentation and their size depends on option 'tabstop'.

Ah, I see now. It wasn't aligned for me either (because I tend to use a tabstop of 2) but I didn't think to fix it. I'll fix it now.

Thanks for pointing me to the memory.c examples - I'll update the comment style.

JackDanger committed:
updating documentation for get_indent_after_label*

JackDanger commented on src/nvim/indent_c.c:
```diff
@@ -504,33 +507,33 @@ get_indent_nolabel (     /* XXX */
+ * @param pp   A pointer to store a reference to the text at this indent.
+ * @returns an integer representing the number of columns of indent
+ *
+ *   label: if (asdf && asdfasdf)
```
@oni-link How's this look?

JackDanger commented:
@oni-link I believe I've addressed all your feedback. How does this PR look now?

oni-link commented:
@JackDanger, see https://github.com/neovim/neovim/pull/1840#commitcomment-9393209. As it is now, `cin_islabel()` checks the cursor line for for a label, but we need to check line `lnum`.

JackDanger commented:
> As it is now, cin_islabel() checks the cursor line for for a label, but we need to check line lnum.

To be clear, you're suggesting there's a bug in the existing implementation and we should modify its behavior?

oni-link commented:
@JackDanger , here is a snippet of the old code.
```c
curwin->w_cursor.lnum = lnum;
l = get_cursor_line_ptr();
 /* XXX */
if (cin_iscase(l, FALSE) || cin_isscopedecl(l) || cin_islabel()) {
```
and here is the first line of `cin_islabel()`
```c
int cin_islabel(void)
{ /* XXX */
  char_u *s = cin_skipcomment(get_cursor_line_ptr());
```
The cursor line in the old code is changed to `lnum`, so that `get_cursor_line_ptr()` in `cin_islabel()` also returns a pointer to line `lnum`. In the new code cursor line and line `lnum` can be different.

aktau commented:
> The cursor line in the old code is changed to lnum, so that get_cursor_line_ptr() in cin_islabel() also returns a pointer to line lnum. In the new code cursor line and line lnum can be different.

Damn, this also mean it will be more difficult to avoid setting/restoring the "cursor pointer".

JackDanger committed:
Passing line number to cin_islabel

JackDanger commented:
@oni-link I'm so glad you noticed that! That slipped right past me. Because `cin_islabel` is only used in 3 places I've simply updated it to accept a line number as an argument. It now relies on global state slightly less.
I've left it in an individual commit here to make it easy to review: https://github.com/JackDanger/neovim/commit/6959e9c0763176d07435c2be5966442ffb5bf443 (though I'd squash and rebase before we would merge this)

oni-link commented:
@JackDanger, the rest of `cin_islabel()` still uses the "cursor line"  instead of `lnum`.
This function is also used in file edit.c, so there would be some more places to update.

Some of the lines in this PR have more than 80 characters (`:set cc=80` to color mark column 80).

JackDanger commented:
@oni-link 
> the rest of cin_islabel() still uses the "cursor line" instead of lnum.

`cin_islabel` stores the `curwin->w_cursor` in a local variable and restores it. I believe setting `curwin->w_cursor.lnum = lnum` immediately after that solves this problem. Do you agree that [this](https://github.com/neovim/neovim/pull/1840/files#diff-c06e34e2b27a2225f245b47f2978516eR288) is a good fix?

> This function is also used in file edit.c, so there would be some more places to update.

Thank you, I'm not sure how I missed that. I've now updated both of the references in `edit.c`.

> Some of the lines in this PR have more than 80 characters (:set cc=80 to color mark column 80).

Thank you for showing me `set cc`, that's new to me! I've fixed the line width throughout the PR.

I've also taken this opportunity to squash the PR into fewer commits for ease of review.

oni-link commented:
> cin_islabel stores the curwin->w_cursor in a local variable and restores it. I believe setting curwin->w_cursor.lnum = lnum immediately after that solves this problem. Do you agree that this is a good fix?

@JackDanger, yes that should fix it.

JackDanger commented:
@oni-link I apologize for the delay in responding to your comments. I've addressed them all and I've rebased this branch off master.
The file `indent_c.c` also now has 2 fewer `clint.py` violations.

justinmk commented:
Sorry this fell through the cracks, but it looks like no one took an interest in finalizing this. I personally don't think we should spend time polishing indent_c, it's never a bottleneck, it's maintained by Bram.
